### PR TITLE
Race condition with received message causes ZMQ_CONNECT_ROUTING_ID to be assigned to wrong socket

### PR DESCRIPTION
**Pull request by mvilim**
Problem: ZMQ_CONNECT_ROUTING_ID can be assigned to incoming socket connection (Issue #3191)

Solution: Add an identifier parameter for local attach to zmq::socket_base_t::attach_pipe

### CODE REVIEW

**Comment by mvilim** [comment]
I followed the pattern of "subscribe_to_all_" and added another parameter to attach_pipe. However, this is a bit ugly because it appears in a lot of socket types where it has no impact (and it makes the default values problematic because we only care about one of the optional parameters in any one place). Let me know if you'd like to try to come up with a different approach.

File: tests/test_connect_rid.cpp [hunk 1]
```diff
@@ -189,13 +189,115 @@ void test_router_2_router (bool named_)
     zmq_ctx_destroy (ctx);
 }
 
+void test_router_2_router_while_receiving ()
+{
+    void *xbind, *zbind, *yconn;
+    int ret;
+    char buff[256];
+    char msg[] = "hi 1";
+    const char *xbindip = "tcp://127.0.0.1:5555";
```
Comment (bluca) [inline tests/test_connect_rid.cpp]: don't hard-code the bind ports, use a wildcard

**Comment by mvilim** [comment]
Fixed. I force-pushed a commit with the fix squashed in. I wasn't sure if that was preferred to having commits with fixes in the pull request.

**Comment by bluca** [comment]
That's fine, thanks

**Comment by bluca** [comment]
@mvilim could you please send a new pull request or an email with a relicense statement? As explained in https://github.com/zeromq/libzmq/blob/master/.github/CONTRIBUTING.md and https://github.com/zeromq/libzmq/tree/master/RELICENSE we are trying to relicense libzmq under MPL-2. Thank you!

**Pull request merged**